File: app/src/main/java/de/lgohlke/pebuild/graph/ExecutionGraph.java
package de.lgohlke.pebuild.graph;

import de.lgohlke.pebuild.JobTriggerHandler;
import de.lgohlke.pebuild.StepExecutor;
import de.lgohlke.pebuild.TimingContext;
import de.lgohlke.pebuild.graph.validators.CycleValidator;
import de.lgohlke.pebuild.graph.validators.ReferencedJobMissingValidator;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

import java.time.Duration;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

@Slf4j
@Deprecated
public class ExecutionGraph implements JobTriggerHandler {

    private final Map<StepExecutor, Collection<StepExecutor>> waitList = new ConcurrentHashMap<>();

    @Getter
    private final Collection<StepExecutor> jobs;
    @Getter
    private Duration timeout;

    private final Map<String, StepExecutor> jobNameMap;
    private final CountDownLatch toBeCompletedLatch;

    private final TransferQueue<TimingContext> timingContextChannel = new LinkedTransferQueue<>();

    private ExecutorService executorService = Executors.newFixedThreadPool(5);

    private ExecutionGraph(Collection<StepExecutor> jobs, Duration timeout) {
        this.jobs = jobs;
        this.timeout = timeout;

        toBeCompletedLatch = new CountDownLatch(jobs.size());
        Map<String, StepExecutor> map = jobs.stream()
                                            .map(j -> new AbstractMap.SimpleEntry<>(j.getName(), j))
                                            .collect(Collectors.toMap(
                                                    AbstractMap.SimpleEntry::getKey,
                                                    AbstractMap.SimpleEntry::getValue));
        jobNameMap = new ConcurrentHashMap<>(map);
    }

    @Override
    public String toString() {
        return jobs.stream()
                   .map(j -> "(" + j + ")")
                   .collect(Collectors.joining("->"));
    }

    @SneakyThrows
    public void execute() {
        createTimingCollector();
        createWaitList();

        synchronized (this) {
            startNextJobs();
        }
        waitAtMaximumTillTimeout();
        shutDown();
    }

    private void createTimingCollector() {
        Callable<Boolean> keepRunning = () -> !(executorService.isTerminated() || executorService.isShutdown());

        executorService.submit(new TimingCollector(timingContextChannel, keepRunning));
    }

    private void createWaitList() {
        waitList.clear();
        jobs.forEach(j -> waitList.put(j, j.getWaitForJobs()));
        log.info("waitList: " + waitList);
    }

    private void shutDown() {
        log.info("shutdown");
        executorService.shutdown();
        try {
            executorService.awaitTermination(10, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            log.error("failed to terminate executorService", e);
        }
    }

    private void waitAtMaximumTillTimeout() throws InterruptedException {
        long milliseconds = timeout.getSeconds() * 1000 + timeout.getNano() / 1_000_000;
        toBeCompletedLatch.await(milliseconds, TimeUnit.MILLISECONDS);
    }

    @Override
    public void onComplete(String jobName, TimingContext timingContext) {
        log.info("oncomplete: " + jobName);
        toBeCompletedLatch.countDown();

        timingContextChannel.offer(timingContext);

        Optional<StepExecutor> jobOptional = Optional.ofNullable(jobNameMap.get(jobName));

        synchronized (this) {
            jobOptional.ifPresent(job -> {
                waitList.forEach((j, waitForJobs) -> {
                    if (waitForJobs.remove(job)) {
                        log.info("removed '" + job + "' from waitlist of " + j);
                    }
                });
            });

            startNextJobs();
        }
    }

    private void startNextJobs() {
        List<StepExecutor> removalFromWaitlist = new ArrayList<>();

        waitList.forEach((j, waitForJobs) -> {
            if (waitForJobs.isEmpty()) {
                executorService.submit(j::execute);
                removalFromWaitlist.add(j);
            }
        });

        removalFromWaitlist.forEach(waitList::remove);
        removalFromWaitlist.clear();
    }

    public static class Builder {
        private final Duration DEFAULT_TIMEOUT = Duration.ofMinutes(10);

        private Collection<StepExecutor> stepExecutors = new HashSet<>();
        private Duration timeout = DEFAULT_TIMEOUT;

        public Builder addJob(StepExecutor executor) {
            if (stepExecutors.add(executor)) {
                return this;
            }
            throw new DuplicateJobException("tried to add duplicated executor: " + executor);
        }

        public Builder timeout(Duration timeout) {
            this.timeout = timeout;
            return this;
        }

        void sort() {
            stepExecutors = new LinkedList<>(TopologicalSorter.sort(stepExecutors));
        }

        void validate() {
            ReferencedJobMissingValidator.validate(stepExecutors);
            CycleValidator.validate(stepExecutors);
        }

        public ExecutionGraph build() {

            validate();
            sort();

            return new ExecutionGraph(stepExecutors, timeout);
        }

        class DuplicateJobException extends RuntimeException {
            DuplicateJobException(String message) {
                super(message);
            }
        }
    }
}

File: app/src/main/kotlin/de/lgohlke/pebuild/graph/ExecutionGraph.kt
package de.lgohlke.pebuild.graph

import de.lgohlke.pebuild.StepExecutor
import de.lgohlke.pebuild.graph.validators.CycleValidator
import de.lgohlke.pebuild.graph.validators.ReferencedJobMissingValidator
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.publisher.toFlux
import reactor.core.scheduler.Schedulers.elastic
import java.time.Duration
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue
import kotlin.collections.HashSet
import kotlin.collections.LinkedHashMap
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.set

class ExecutionGraph(val steps: Collection<StepExecutor>,
                     private val finalStep: StepExecutor,
                     val timeout: Duration) {

    fun execute() {
        val timingContext = TimingContext()
        timingContext.time("__global__") { innerExecute(timingContext) }
        timingContext.save()
    }

    private fun innerExecute(timingContext: TimingContext) {
        val cachedSteps = steps.map { step -> step to createCachedMono(step, timingContext) }.toMap()
        val graph = createExecutionGraph(finalStep, cachedSteps).log()

        graph.timeout(timeout)
            // fail fast
            // stop the whole graph in case of timeout or another error
            .onErrorResume { Mono.empty() }
            .blockLast()
    }

    class TimingContext {
        data class Timing(val start: Long,
                          val end: Long)

        private val timings = ConcurrentHashMap<String, Long>()
        private val finishedTimings = ConcurrentLinkedQueue<Timing>()

        private fun start(name: String) {
            timings[name] = System.currentTimeMillis()
        }

        private fun stop(name: String) {
            val start = timings[name] ?: throw IllegalStateException("could not stop a not started task: $name")
            val end = System.currentTimeMillis()

            finishedTimings.add(Timing(start, end))
            timings.remove(name)
        }

        fun save() {
//            finishedTimings.toArray()
            // TODO
        }

        fun time(name: String,
                 execution: () -> Unit) {
            start(name)
            try {
                execution()
            } finally {
                stop(name)
            }
        }
    }

    // TODO
    class StepError(val step: StepExecutor,
                    private val error: Throwable) : java.lang.RuntimeException(error)

    private fun createCachedMono(step: StepExecutor,
                                 timingContext: TimingContext): Mono<StepExecutor> {
        val runnable = Runnable {
            timingContext.time(step.name) { step.execute() }
        }
        return Mono.fromRunnable<StepExecutor>(runnable)
            .name(step.name)
            .subscribeOn(elastic())
            .timeout(step.timeout)
            .onErrorMap { error -> StepError(step, error) }
            .doOnError { step.cancel() }
            .cache()
            .doOnCancel { step.cancel() }
    }

    private fun createExecutionGraph(step: StepExecutor,
                                     cachedNodes: Map<StepExecutor, Mono<StepExecutor>>): Flux<StepExecutor> {
        val thisNode = cachedNodes[step]?.toFlux() ?: throw IllegalStateException("missing step in cached map: $step")
        val preFluxes = step.waitForJobs.map { predecessor -> createExecutionGraph(predecessor, cachedNodes) }
        val merge = Flux.merge(Flux.fromIterable(preFluxes))
        return Flux.concat(merge, thisNode)
    }

    class Builder {

        private val steps = HashSet<StepExecutor>()
        private var timeout = Duration.ofDays(999)

        fun addJob(executor: StepExecutor): Builder {
            if (!steps.add(executor)) {
                throw DuplicateJobException("$executor is already added")
            }
            return this
        }

        private fun createFinalStep(executors: Collection<StepExecutor>): StepExecutor {
            val finalSteps = createSuccessorMap(executors)
                .filter { (_, successors) -> successors.isEmpty() }
                .map { (node, _) -> node }
            val theFinalStep = object : StepExecutor("end", "") {}
            finalSteps.forEach { step -> theFinalStep.waitFor(step) }
            return theFinalStep
        }

        private fun createSuccessorMap(steps: Collection<StepExecutor>): Map<StepExecutor, Set<StepExecutor>> {
            val successorMap = LinkedHashMap<StepExecutor, MutableSet<StepExecutor>>()

            steps.forEach { step ->
                successorMap[step] = LinkedHashSet()
            }

            steps.forEach { step ->
                step.waitForJobs.forEach { s ->
                    successorMap[s]?.add(step)
                }
            }
            return successorMap
        }

        fun timeout(timeout: Duration): Builder {
            this.timeout = timeout
            return this
        }

        fun addJobs(vararg jobs: StepExecutor): Builder {
            jobs.forEach { addJob(it) }
            return this
        }

        private fun validate() {
            ReferencedJobMissingValidator.validate(steps)
            CycleValidator.validate(steps)
        }

        fun build(): ExecutionGraph {
            validate()
            val finalStep = createFinalStep(steps)
            steps.add(finalStep)
            return ExecutionGraph(steps, finalStep, timeout)
        }

        class DuplicateJobException(message: String) : RuntimeException(message)
    }
}